src/os_common/os_fhandle.c

```diff
@@ -355,6 +370,117 @@ __wt_close(WT_SESSION_IMPL *session, WT_FH **fhp)
 	return (__handle_close(session, fh, true));
 }
 
+/*
+ * __wt_fsync_background_chk --
+ *	Return if background fsync is supported.
+ */
+int
+__wt_fsync_background_chk(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_FH *fh, *fhnext;
+	WT_FILE_HANDLE *handle;
+
+	conn = S2C(session);
+	WT_ASSERT(session, !F_ISSET(conn, WT_CONN_READONLY));
+	__wt_spin_lock(session, &conn->fh_lock);
+	/*
+	 * Look for the first non-WiredTiger file handle and see if
+	 * the fsync nowait function is supported.
+	 */
+	TAILQ_FOREACH_SAFE(fh, &conn->fhqh, q, fhnext) {
+		handle = fh->handle;
+		/* Skip over WiredTiger owned files. */
+		if (F_ISSET(fh, WT_FH_WIREDTIGER_OWNED))
+			continue;
+		/*
+		 * If we don't have a function, return WT_NOTFOUND,
+		 * otherwise return 0. In any case, we are done with
+		 * the loop.
+		 */
+		if (handle->fh_sync_nowait == NULL)
+			ret = WT_NOTFOUND;
+		break;
+	}
+	__wt_spin_unlock(session, &conn->fh_lock);
+	return (ret);
+}
+
+/*
+ * __wt_fsync_background --
+ *	POSIX background fsync for all dirty file handles.
+ */
+int
+__wt_fsync_background(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_FH *fh, *fhnext;
+	WT_FILE_HANDLE *handle;
+	uint64_t now;
+
+	conn = S2C(session);
+	WT_ASSERT(session, !F_ISSET(conn, WT_CONN_READONLY));
+	__wt_spin_lock(session, &conn->fh_lock);
+	TAILQ_FOREACH_SAFE(fh, &conn->fhqh, q, fhnext) {
+		handle = fh->handle;
+		WT_STAT_CONN_INCR(session, fsync_all_fh_total);
+		if (handle->fh_sync_nowait == NULL ||
+		    fh->written < WT_CAPACITY_FILE_THRESHOLD)
+			continue;
+		/* Skip over WiredTiger owned files. */
+		if (F_ISSET(fh, WT_FH_WIREDTIGER_OWNED))
+			continue;
+		now = __wt_clock(session);
+		if (fh->last_sync == 0 ||
+		    WT_CLOCKDIFF_SEC(now, fh->last_sync) > 0) {
+			/*
+			 * Adjust our next handle to skip over any that are
+			 * owned by WiredTiger. Adding a reference to those
+			 * interferes with other internal operations such as
+			 * log archiving.
+			 */
+			while (fhnext != NULL &&
+			    F_ISSET(fhnext, WT_FH_WIREDTIGER_OWNED))
+				fhnext = TAILQ_NEXT(fhnext, q);
+			/*
+			 * Increment our ref count on the current and next
+			 * handle. That way both are guaranteed valid when we
+			 * lock again after the fsync.
+			 */
+			++fh->ref;
+			if (fhnext != NULL)
+				++fhnext->ref;
+			__wt_spin_unlock(session, &conn->fh_lock);
+			ret = __wt_fsync(session, fh, false);
+			/*
+			 * Although we send in the false flag to indicate a
+			 * non-blocking background fsync, there is no guarantee
+			 * that it doesn't block. If we wanted to detect if it
+			 * is blocking, adding a clock call and checking the
+			 * time would be done here.
+			 */
+			/*
+			 * Even if we get an error, we still need to lock and
+			 * decrement our reference counts.
+			 */
+			if (ret == 0) {
+				WT_STAT_CONN_INCR(session, fsync_all_fh);
+				fh->last_sync = now;
+				WT_PUBLISH(fh->written, 0);
```

Comment:
Did you consider having two values in the `WT_FH` structure, one for the written bytes and one for the point at which fsync wrote the bytes? That should mean less collisions, given that the background sync thread is the only thread writing one of those values.